### PR TITLE
feat: allow presets to import modules

### PR DESCRIPTION
**Pull request by SomaticIT**
Hello,

Thank you for building this library. The concept is very interesting.

This PR tries to allow more complex presets to be written.

### Motivation

I try to use this project to create a cli which helps my developers scaffoling microservices projects.
In this context, I built a lot of presets to initialize a project, add components, tooling, etc.

In this "kind of" preset collection, we have a lot of shared logic which can be refactored but it's impossible to import a module inside the `vm` environment.

### Solution

I saw that you use `esbuild` to be able to accept TypeScript code so I decided to try using it to bundle imported dependencies.
I also had to create a `vm` context which better emulate node env to allow (simple) external `npm` modules to work.
With some configuration, it works for a lot of simple cases.

So to summarize, this PR:
1. Extends the `vm` context to match more closely to node env
2. Use `esbuild` to bundle dependencies when `import` statement is detected

### Limitations

This implementation has some limitations (less than before but still):
- Many npm modules will not work in a `vm` environment
- Many npm modules are incompatible with `esbuild`

Despite this, it is very useful for local imports and it works with some npm deps.

### Review

If you look at the `createContext` method, I commented some globals that should be available in `node` >= 12 but are not available in TypeScript in the specific configuration of this project.

I see two choices to enable these globals:
1. Include `DOM` lib in `tsconfig.json`, which is not a good thing because it allows other DOM globals to be incorrectly used in the project.
2. Add an ambiant `globals.d.ts` in the project to define missing globals (even as any)

What do you think?

### CODE REVIEW

File: src/Importer/ModuleImporter.ts
Comment (innocenzi) [inline src/Importer/ModuleImporter.ts]: Is it useful to use `transformSync` instead of `buildSync` when there is no `import` statement? If a user want to update a file to add an import line, `import ` will be detected and `buildSync` will be used, even if the actual preset doesn't import anything. Would that be an issue?

File: src/Importer/ModuleImporter.ts
Comment (innocenzi) [inline src/Importer/ModuleImporter.ts]: > Add an ambiant globals.d.ts in the project to define missing globals (even as any)

Did you test this? Wouldn't that be an issue if the global does not actually exist? I don't think we can partially include `dom.d.ts`. 

Also, how are you using these in your presets?

File: src/Importer/ModuleImporter.ts
Comment (SomaticIT) [reply src/Importer/ModuleImporter.ts]: Thank you for this review.

I added this test to reduce at its minimum the impact on existing presets.

IMHO, `buildSync` should be similar to `transformSync` when there is no `import` statement so it should work with existing presets but I wanted to be sure that I didn't introduce a breaking change.

If you want, I can try to remove this test so you can test if it works well in your existing presets.
Do you agree?

Comment (innocenzi) [reply src/Importer/ModuleImporter.ts]: I also think they would be similar. I don't think it's a breaking change and I'd rather have consistency over all types of preset. I'd be happy to test my presets after you make the change!

File: src/Importer/ModuleImporter.ts
Comment (SomaticIT) [reply src/Importer/ModuleImporter.ts]: > Did you test this? Wouldn't that be an issue if the global does not actually exist?

Adding an ambiant globals.d.ts will work on the TypeScript side but it has no effect on javascript side. So if the global does not exist, an error will be thrown.

All variables I use here are available in node >= 12 so it's pretty safe on these versions. This project has the following engine `"node": ">=12"` declared in its package.json so I considered it safe.

To be safer, I tried a new version that will work on previous node versions. It also avoid adding an ambiant global.d.ts.
Is it OK for you?

> I don't think we can partially include dom.d.ts.

I agree, I think using the above solution is better.

> Also, how are you using these in your presets?

I'm using only `Buffer` for some custom file manipulations or `console` for some tests. But when I tried to import some npm modules, I had many troubles because of missing globals (mostly `process`, `console` and `Buffer`).

I concluded that we should better virtualize node environment to increase compatibility with npm modules.
To achieve this, I added all module and node globals to vm context.

File: src/Importer/ModuleImporter.ts
Comment (SomaticIT) [reply src/Importer/ModuleImporter.ts]: I did the change.

**Comment by SomaticIT** [comment]
Hi @innocenzi,

Did you have some time to test this on your projects?
Do you want me to do some other changes before merging?

File: src/Importer/ModuleImporter.ts
Comment (innocenzi) [inline src/Importer/ModuleImporter.ts:162]: I don't understand this. `global` will never be something else than `NodeJS.Global`, right? So, `TextDecorator`, `TextEncoder`, `URL`, `URLSearchParams` and `WebAssembly` will always be `undefined` inside the context, thus not solving the issue?

**Comment by innocenzi** [comment]
Sorry, didn't really have time this week. I left a comment, I'm down to merge this once I understand the purpose of your last change.

File: src/Importer/ModuleImporter.ts
Comment (SomaticIT) [reply src/Importer/ModuleImporter.ts:162]: These 5 globals are not defined in `NodeJS.Global` but it's only a TypeScript issue.
They are available in `nodejs >= 12`. You can test it on your computer using a simple repl.

```bash
$ node
Welcome to Node.js v12.22.1.
Type ".help" for more information.
> global.TextDecoder
[class TextDecoder]
>
```

I added the `global.[...]` to ensure that it will not throw on previous versions of node. If a global is unavailable on a specific version it will fallback to `undefined` (as expected).

**Comment by SomaticIT** [comment]
There is absolutely no problem 😁 !

I answered your question. Hope it's clear.

File: src/Importer/ModuleImporter.ts
Comment (innocenzi) [reply src/Importer/ModuleImporter.ts:162]: Makes sense, thanks for the clarification!

**Review by innocenzi — APPROVED** [review]

**Comment by innocenzi** [comment]
Perfect, thank you! Tagged a new release: https://github.com/preset/cli/releases/tag/v0.2.14.

**Comment by SomaticIT** [comment]
👍 Thanks !

**Pull request merged**
